### PR TITLE
Add support for semicolon stripping to DbApiHook, PrestoHook, and TrinoHook

### LINKED ISSUE #34828: Trino SQL split trailing semi-colon
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

We're currently using Airflow 2.6.1 but I believe the issue persists in the latest versions.

The SQLExecuteQueryOperator has the functionality to split_statements using a semi-colon. Trino is one of the few execution engines that does not accept a trailing semi-colon. sqlparse leaves the semi-colon after splitting.

The [TrinoHook](https://github.com/apache/airflow/blob/main/airflow/providers/trino/hooks/trino.py#L73) doesn't implement it's own run method so this is then handled by the DbApiHook.
split_statements on the DbApiHook uses [sqlparse](https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/hooks/sql.py#L278) to do the splitting.

### What you think should happen instead

When passing in multiple SQL statements that are semi-colon seperated, they should be split without keeping the semi-colon so that Trino considers it to be valid.

Because of Trino's current [implementation](https://github.com/trinodb/trino/issues/18472) if we want to make use of this functionality I believe that custom splitting logic should be implemented in Trino's hook.
If the Trino implementation is revised later on so that it accepts semi-colons then this could be removed again.

### How to reproduce

Run the SQLExecuteQueryOperator using Trino with multiple SQL statements that are semi-colon seperated.
E.g. 

```
SQLExecuteQueryOperator(
        task_id="semi_colon_test",
        conn_id="trino_default",
        sql="SELECT 1; SELECT 2",
        split_statements=True,
        handler=list,
    )
```

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==8.0.0
apache-airflow-providers-celery==3.1.0
apache-airflow-providers-common-sql==1.4.0
apache-airflow-providers-databricks==4.1.0
apache-airflow-providers-ftp==3.3.1
apache-airflow-providers-http==4.3.0
apache-airflow-providers-imap==3.1.1
apache-airflow-providers-mysql==5.0.0
apache-airflow-providers-postgres==5.4.0
apache-airflow-providers-presto==5.0.0
apache-airflow-providers-redis==3.1.0
apache-airflow-providers-sqlite==3.3.2
apache-airflow-providers-trino==5.0.0

### Deployment

Other

### Deployment details

Self hosted in AWS using ECS Fargate, RDS (Postgres) and Elasticache (Redis).

### Anything else

Error message

`trino.exceptions.TrinoUserError: TrinoUserError(type=USER_ERROR, name=SYNTAX_ERROR, message="line 1:84: mismatched input ';'. Expecting: '.', <EOF>", query_id=20231009_081239_17877_i2yca)`

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by Illumaria**
Closes: #34828

### CODE REVIEW

**Comment by potiuk** [comment]
You will have to:

1) bump common.sql minor version (new feature) in airfow/providers/common/sql/provider.yaml
2) In all modified providers you will have to update dependency in provider yaml to be:  `common-sql >= THE(_NEW_COMMON_SQL_VERSION`

**Comment by Illumaria** [comment]
Whoops... Sorry. 🙏🏻

File: airflow/providers/common/sql/provider.yaml
Comment (eladkal) [inline airflow/providers/common/sql/provider.yaml]: Please remove this entry

File: airflow/providers/presto/provider.yaml
Comment (eladkal) [inline airflow/providers/presto/provider.yaml]: Change to 1.17.0

**Review by eladkal — CHANGES_REQUESTED** [review]

File: airflow/providers/presto/provider.yaml
Comment (potiuk) [reply airflow/providers/presto/provider.yaml]: Yes. The version has been bumped already by #40751 and it would have to be 1.17.0 anyway as this is a new feature (MINOR is the middle number - the last one is PATCHLEVEL)

**Comment by eladkal** [comment]
@Illumaria can you address the needed changes? I'm hoping to include this PR in the next release

File: airflow/providers/common/sql/provider.yaml
Comment (eladkal) [inline airflow/providers/common/sql/provider.yaml]:
```suggestion
```

File: airflow/providers/presto/provider.yaml
Comment (eladkal) [inline airflow/providers/presto/provider.yaml]:
```suggestion
  - apache-airflow-providers-common-sql>=1.17.0
```

File: airflow/providers/trino/provider.yaml
Comment (eladkal) [inline airflow/providers/trino/provider.yaml]:
```suggestion
  - apache-airflow-providers-common-sql>=1.17.0
```

File: generated/provider_dependencies.json
Comment (eladkal) [inline generated/provider_dependencies.json]:
```suggestion
      "apache-airflow-providers-common-sql>=1.17.0",
```

File: generated/provider_dependencies.json
Comment (eladkal) [inline generated/provider_dependencies.json]:
```suggestion
      "apache-airflow-providers-common-sql>=1.17.0",
```

**Comment by eladkal** [comment]
needs rebase and resolve conflicts

**Comment by Illumaria** [comment]
@eladkal Sure, I’ll do it by the end of the weekend.

**Comment by potiuk** [comment]
> @eladkal Hi! Could you please help me understand what I should do to fix the failing tests? :(

There are regular unit tests failing and they should be fixed. For compatibility tests - when they fail, they point you with the link to the documentation that desribes possible approaches and shows some examples how to deal with compatibility tests, and there are a number of examples already implemented in the code:

https://github.com/apache/airflow/blob/main/contributing-docs/testing/unit_tests.rst#compatibility-provider-unit-tests-against-older-airflow-releases

**Comment by eladkal** [comment]
> @eladkal Hi! Could you please help me understand what I should do to fix the failing tests? :(

`
=========================== short test summary info ============================
FAILED tests/providers/google/cloud/openlineage/test_mixins.py::TestBigQueryOpenLineageMixin::test_bq_job_information - TypeError: split() got an unexpected keyword argument 'strip_semicolon'
FAILED tests/providers/google/cloud/openlineage/test_mixins.py::TestBigQueryOpenLineageMixin::test_bq_script_job_information - TypeError: split() got an unexpected keyword argument 'strip_semicolon'
FAILED tests/providers/google/cloud/operators/test_bigquery.py::TestBigQueryInsertJobOperator::test_execute_openlineage_events - TypeError: split() got an unexpected keyword argument 'strip_semicolon'
FAILED tests/providers/google/cloud/operators/test_bigquery.py::TestBigQueryInsertJobOperator::test_execute_fails_openlineage_events - TypeError: split() got an unexpected keyword argument 'strip_semicolon'
============ 4 failed, 4159 passed, 63 skipped in 271.27s (0:04:31) ============
`

**Comment by eladkal** [comment]
@Illumaria were you able to work on fixing the test?

**Comment by Illumaria** [comment]
> @Illumaria were you able to work on fixing the test?

No, not yet. :( I’m away from home now, will be back on September 30th.

Good news is I kinda need these changes for my work, so I won’t leave this request unfinished.

**Comment by eladkal** [comment]
Tests are failing :(

**Comment by Illumaria** [comment]
> Tests are failing :(

Thank you, I can clearly see that. :/ Is there any way to run those failing tests locally? So far I've tried
```
breeze --integration all
```
and it's all green, and
```
breeze testing non-db-tests
```
for which I get these errors (they are also very confusing):
image

File: providers/src/airflow/providers/openlineage/provider.yaml
Comment (eladkal) [inline providers/src/airflow/providers/openlineage/provider.yaml]:
```suggestion
  - apache-airflow-providers-common-sql>=1.20.0
```

We need to bump the versions again

Comment (Illumaria) [reply providers/src/airflow/providers/openlineage/provider.yaml]: Hopefully, this is done.

**Comment by eladkal** [comment]
I think for this to actually work you will need to specify directly 1.20.0 in
https://github.com/apache/airflow/blob/845a16057625127490ba8d655b3fe1114373ff7b/providers/src/airflow/providers/common/sql/provider.yaml#L27-L28

otherwise i am not sure how the CI can build the min version required by the other provider dependencies?

**Comment by Illumaria** [comment]
> I think for this to actually work you will need to specify directly 1.20.0 in
> 
> https://github.com/apache/airflow/blob/845a16057625127490ba8d655b3fe1114373ff7b/providers/src/airflow/providers/common/sql/provider.yaml#L27-L28
> 
> otherwise i am not sure how the CI can build the min version required by the other provider dependencies? but need to verify with @potiuk on this one

Yeah but I thought I'm not supposed to touch that, as it says in https://github.com/apache/airflow/blob/845a16057625127490ba8d655b3fe1114373ff7b/providers/src/airflow/providers/common/sql/provider.yaml#L26:

> # note that those versions are maintained by release manager - do not update them manually

**Review by eladkal — APPROVED** [review]

**Comment by eladkal** [comment]
failures :(
```

=========================== short test summary info ============================
FAILED providers/tests/presto/hooks/test_presto.py::TestPrestoHook::test_split_sql_string - AssertionError: assert equals failed
  [               [              
    'SELECT 1',     'SELECT 1;', 
    'SELECT 2',     'SELECT 2',  
  ]               ]
FAILED providers/tests/trino/hooks/test_trino.py::TestTrinoHook::test_split_sql_string - AssertionError: assert equals failed
  [               [              
    'SELECT 1',     'SELECT 1;', 
    'SELECT 2',     'SELECT 2',  
  ]               ]
= 2 failed, 12280 passed, 9740 skipped, 2 xfailed, 5 warnings in 616.06s (0:10:16) =
```

**Comment by Illumaria** [comment]
@eladkal Yeah I see. :(

Do I get it correctly that due to the following change
[providers/src/airflow/providers/openlineage/sqlparser.py](https://github.com/apache/airflow/pull/41916/files#diff-7bbec90955ebce43e1a3996d34f3565ff3f3c3f72244c8a6a47bd35c5a890acb)
I have to bump `openlineage` provider version as well?

**Comment by eladkal** [comment]
> I have to bump openlineage provider version as well?

Let try. Add 1.14.0 entry in openlineage provider.yaml

**Comment by mobuchowski** [comment]
@Illumaria I don't think the test failures are related to OL in any way, it's just you're calling method with default argument, in test do not pass the optional argument, and then it fails - because it uses default `False` argument. It does not pick up class member from `TrinoHook` or `PrestoHook` because nothing actually in the class `split_sql_string` refers to them.

**Comment by Illumaria** [comment]
@mobuchowski OMG, you're right. I completely overlooked that. 🤦🏻‍♂️ Thanks!

**Comment by mobuchowski** [comment]
🙌 

Now, looking at rest of the tests, I think there is additional compatibility issue,

**Comment by eladkal** [comment]
> 🙌
> 
> Now, looking at rest of the tests, I think there is additional compatibility issue - you might need to bump required version of common.sql provider for all of the SQL providers

Yeah lets bump for all providers that are using it

**Comment by eladkal** [comment]
@Illumaria i am hoping to cut a release today. Let me know if you have time to finish this one

**Comment by Illumaria** [comment]
@eladkal I'm trying. :(

**Comment by Illumaria** [comment]
Ugh... I don't get it. The failing test reports aren't clear on how to fix them. =_=

**Comment by eladkal** [comment]
I triggered a full rebuild of the test suite lets see how it goes

**Comment by eladkal** [comment]
> Ugh... I don't get it. The failing test reports aren't clear on how to fix them. =_=

The issues are dependency conflict around standard provider, airflow 3 and sql provider in Python 3.9
So possibly this will be resolved in https://github.com/apache/airflow/pull/43556

**Comment by eladkal** [comment]
```
Requirement already satisfied: apache-airflow-providers-sqlite in /opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages (from apache-airflow==3.0.0.dev0) (3.9.0)
INFO: pip is looking at multiple versions of apache-airflow-providers-standard to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install apache-airflow-providers-standard==0.0.1 and apache-airflow==3.0.0.dev0 because these package versions have conflicting dependencies.

The conflict is caused by:
    apache-airflow 3.0.0.dev0 depends on apache-airflow-providers-common-sql
    apache-airflow-providers-standard 0.0.1 depends on apache-airflow-providers-common-sql>=1.20.0
```

@potiuk should we put this one on hold till #43556 is resolved?

**Comment by eladkal** [comment]
@Illumaria Please add `common-sql` to `CHICKEN_EGG_PROVIDERS`. That should solve the issue.
https://github.com/apache/airflow/blob/da502427b7204fac5f34ecf138aada37530258c2/dev/breeze/src/airflow_breeze/global_constants.py#L570

**Comment by eladkal** [comment]
> @Illumaria Please add `common-sql` to `CHICKEN_EGG_PROVIDERS`. That should solve the issue.
> 
> https://github.com/apache/airflow/blob/da502427b7204fac5f34ecf138aada37530258c2/dev/breeze/src/airflow_breeze/global_constants.py#L570

Added it in https://github.com/apache/airflow/pull/41916/commits/e36974ae7da7f3f5102c6a11c6febf63a6ed1475 lets hope it  will give us green build

File: dev/breeze/src/airflow_breeze/global_constants.py
Comment (eladkal) [inline dev/breeze/src/airflow_breeze/global_constants.py]:
```suggestion
CHICKEN_EGG_PROVIDERS = " ".join(["standard amazon common.sql"])
```

**Comment by potiuk** [comment]
Rebased after merging #43913

**Comment by eladkal** [comment]
OK @Illumaria I think we sort out all the issues related to breeze testing. We now see failures which seems related to the PR itself for the back compat tests with Airflow 2.8 can you take a look?

```
=========================== short test summary info ============================
FAILED providers/tests/google/cloud/openlineage/test_mixins.py::TestBigQueryOpenLineageMixin::test_bq_job_information - TypeError: split() got an unexpected keyword argument 'strip_semicolon'
FAILED providers/tests/google/cloud/openlineage/test_mixins.py::TestBigQueryOpenLineageMixin::test_bq_script_job_information - TypeError: split() got an unexpected keyword argument 'strip_semicolon'
FAILED providers/tests/google/cloud/operators/test_bigquery.py::TestBigQueryInsertJobOperator::test_execute_openlineage_events - TypeError: split() got an unexpected keyword argument 'strip_semicolon'
FAILED providers/tests/google/cloud/operators/test_bigquery.py::TestBigQueryInsertJobOperator::test_execute_fails_openlineage_events - TypeError: split() got an unexpected keyword argument 'strip_semicolon'
====== 4 failed, 4199 passed, 79 skipped, 1 warning in 261.82s (0:04:21) =======
```

**Comment by Illumaria** [comment]
Uhm... Is this some network issue? O_o

```
Run docker run -v "${GITHUB_WORKSPACE}:/workspace" -u 0:0 bash -c "rm -rf /workspace/*"
Unable to find image 'bash:latest' locally
docker: Error response from daemon: Head "https://registry-1.docker.io/v2/library/bash/manifests/latest": Get "https://auth.docker.io/token?account=githubactions&scope=repository%3Alibrary%2Fbash%3Apull&service=registry.docker.io": net/http: request canceled (Client.Timeout exceeded while awaiting headers).
See 'docker run --help'.
Error: Process completed with exit code [12](https://github.com/apache/airflow/actions/runs/11816640286/job/32920825020?pr=41916#step:2:13)5.
```

**Comment by potiuk** [comment]
> Uhm... Is this a network issue on GitHub side? O_o

Yeah.... Happens

**Comment by potiuk** [comment]
We can re-start that one failed job later after this PR completes all other checks.

**Review by vgvoleg — APPROVED** [review]

**Comment by Illumaria** [comment]
@potiuk Could you please re-run the failed test job? I don't think I have the rights.

**Review by potiuk — APPROVED** [review]
NICE! finally!

**Comment by potiuk** [comment]
Wooooohooo! (CC: @eladkal )

**Comment by potiuk** [comment]
Thanks @Illumaria for the patience. It was not an easy one and it uncovered a number of edge-cases in our CI that we had to fix :)


#protm

**Comment by eladkal** [comment]
Great work @Illumaria 👏🏻👏🏻

**Comment by willshen99** [comment]
The PR seems not included in 2.10.5 release. Do we have an estimate when and what version this PR will be in a release?

**Pull request merged**
